test_runner/regress/test_tenants_with_remote_storage.py

```diff
@@ -189,3 +193,123 @@ def expect_tenant_to_download_timeline(
             ), f"Tenant {tenant_id} should have no downloads in progress"
             return
     assert False, f"Tenant {tenant_id} is missing on pageserver"
+
+
+@pytest.mark.parametrize("remote_storage_kind", [RemoteStorageKind.LOCAL_FS])
+def test_tenant_upgrades_index_json_from_v0(
+    neon_env_builder: NeonEnvBuilder, remote_storage_kind: RemoteStorageKind
+):
+    # getting a too eager compaction happening for this test would not play
+    # well with the strict assertions.
+    neon_env_builder.pageserver_config_override = "tenant_config.compaction_period='1h'"
+
+    neon_env_builder.enable_remote_storage(remote_storage_kind, "test_tenant_upgrades_index_json")
+
+    # launch pageserver, populate the default tenants timeline, wait for it to be uploaded,
+    # then go ahead and modify the "remote" version as if it was downgraded, needing upgrade
+    env = neon_env_builder.init_start()
+    pageserver_http = env.pageserver.http_client()
+    pg = env.postgres.create_start("main")
+
+    tenant_id = TenantId(pg.safe_psql("show neon.tenant_id")[0][0])
+    timeline_id = TimelineId(pg.safe_psql("show neon.timeline_id")[0][0])
+
+    with pg.cursor() as cur:
+        cur.execute("CREATE TABLE t0 AS VALUES (123, 'second column as text');")
+        current_lsn = Lsn(query_scalar(cur, "SELECT pg_current_wal_flush_lsn()"))
+
+    # flush, wait until in remote storage
+    wait_for_last_record_lsn(pageserver_http, tenant_id, timeline_id, current_lsn)
+    pageserver_http.timeline_checkpoint(tenant_id, timeline_id)
+    wait_for_upload(pageserver_http, tenant_id, timeline_id, current_lsn)
+
+    env.postgres.stop_all()
+    env.pageserver.stop()
+
+    # remove all local data for the tenant to force redownloading and subsequent upgrade
+    shutil.rmtree(Path(env.repo_dir) / "tenants" / str(tenant_id))
+
+    # downgrade the remote file
+    assert isinstance(env.remote_storage, LocalFsStorage)
+    timeline_path = (
+        env.remote_storage.root
+        / "tenants"
+        / str(tenant_id)
+        / "timelines"
+        / str(timeline_id)
+        / "index_part.json"
+    )
+    with open(timeline_path, "r+") as timeline_file:
+        # keep the deserialized for later inspection
+        orig_index_part = json.load(timeline_file)
+
+        # v1 additions
+        assert "version" in orig_index_part, "index_part.version was added as a debugging aid"
+        assert "layer_metadata" in orig_index_part, "index_part.layer_metadata has the metadata for the layer files"
+
+        v0_keys = {"timeline_layers", "missing_layers", "disk_consistent_lsn", "metadata_bytes"}
+        v0_index_part = { key: orig_index_part[key] for key in v0_keys }
+
+        timeline_file.seek(0)
+        json.dump(v0_index_part, timeline_file)
```

Comment:
Responding to your main review comment here, which I think it was meant to comment on:

> I would move the hardcoded format changes from Rust code (that's expected to change a lot) to e2e test (that's not expected to change much: we're supporting two formats for a few months at least), hardcode that json as string there and replace it in storage, that would be also more readable.

Don't the currently unsolved (IIUC) initdb troubles create a situation where following runs of this test might end up creating the two layer files with different lsn ranges, making it nearly similar replacement effort to use the old "image"?

Agreed on the comment otherwise, unsure how to implement this.

Reply:
Agreed. Will type up a version with v0 image with replaced `timeline_layers`.

Reply:
Is https://github.com/neondatabase/neon/pull/2582/commits/0bf205c53703ff03cde04695bdb50440a8623fd6 good enough?

Reply:
I would imagine so: it removes all scary hardcoded bytes from rust code and puts them into the e2e test, looks good.